keps/sig-api-machinery/4192-svm-in-tree/README.md

```diff
@@ -125,65 +126,111 @@ As an end user using encryption at rest, whenever the key change is detected we
+      // is created. It can avoid races.
     }
 
     // The names of the group, the version, and the resource.
     type GroupVersionResource struct {
-        // The name of the group.
-        Group string `json:"group,omitempty"`
-        // The name of the version.
-        Version string `json:"version,omitempty"`
-        // The name of the resource.
-        Resource string `json:"resource,omitempty"`
+      // The name of the group.
+      Group string `json:"group,omitempty" protobuf:"bytes,1,opt,name=group"`
+      // The name of the version.
+      Version string `json:"version,omitempty" protobuf:"bytes,2,opt,name=version"`
+      // The name of the resource.
+      Resource string `json:"resource,omitempty" protobuf:"bytes,3,opt,name=resource"`
     }
-    
+
+    type MigrationConditionType string
+
+    const (
+      // Indicates that the migration is running.
+      MigrationRunning MigrationConditionType = "Running"
+      // Indicates that the migration has completed successfully.
+      MigrationSucceeded MigrationConditionType = "Succeeded"
+      // Indicates that the migration has failed.
+      MigrationFailed MigrationConditionType = "Failed"
+    )
+
+    // Describes the state of a migration at a certain point.
+    type MigrationCondition struct {
+      // Type of the condition.
+      Type MigrationConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=MigrationConditionType"`
+      // Status of the condition, one of True, False, Unknown.
+      Status corev1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=k8s.io/api/core/v1.ConditionStatus"`
+      // The last time this condition was updated.
+      // +optional
+      LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty" protobuf:"bytes,3,opt,name=lastUpdateTime"`
+      // The reason for the condition's last transition.
+      // +optional
+      Reason string `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+      // A human readable message indicating details about the transition.
+      // +optional
+      Message string `json:"message,omitempty" protobuf:"bytes,5,opt,name=message"`
+    }
+
     // Status of the storage version migration.
     type StorageVersionMigrationStatus struct {
-        // The latest available observations of the migration's current state.
-        // +optional
-        // +patchMergeKey=type
-        // +patchStrategy=merge
-        Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+      // The latest available observations of the migration's current state.
+      // +patchMergeKey=type
+      // +patchStrategy=merge
+      // +listType=map
+      // +listMapKey=type
+      // +optional
+      Conditions []MigrationCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
```

Comment:
during API review, we should look to see if this diffs identically for serialization with metav1.Condition.  If it does, the swap now will help.